### ISSUE TITLE
VSCode is missing code-tunnel binary

### ISSUE THREAD

**Issue by StarLederer** [NixOS/nixpkgs#250794]
### Describe the bug

VSCode has recently added [new CLI capabilities](https://code.visualstudio.com/docs/editor/command-line#_create-remote-tunnel). The `code tunnel` command is currently broken in VSCode from Nixpkgs (v.1.81.1).

### Steps To Reproduce

1. `NIXPKGS_ALLOW_UNFREE=1 nix shell --impure nixpkgs#vscode`
1. `code tunel -help`
1. See the error

### Expected behavior

`code tunnel -help` is expected to produce the same output as documented [here](https://code.visualstudio.com/docs/editor/command-line#_create-remote-tunnel)

### Screenshots

![image](https://github.com/NixOS/nixpkgs/assets/55361752/bdca87b5-93f9-44f2-bec8-d8557710b307)

### Notify maintainers

@LudovicoPiero (I used the git method, and you were the last author, please let me know who else I should @)

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.46, NixOS, 23.11 (Tapir), 23.11.20230820.0b07d49`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

**Comment by LudovicoPiero** [comment]
Well, I only updated the C++ extensions. According to the meta attributes, here's the list of maintainers: @eadwu @synthetica @maxeaubrey @bobby285271 @Enzime

**Comment by bobby285271** [comment]
Ugh, I am thinking of #247921, but also saw https://github.com/microsoft/vscode/issues/190274 that is on 1.81.1 :joy:

**Comment by bobby285271** [comment]
Fetching an insider build I can see the `code-tunnel-insiders` binary, so by "[next stable](https://github.com/microsoft/vscode/issues/189383#issuecomment-1666680413)" it will likely be 1.82 :joy:

**Comment by bobby285271** [comment]
@StarLederer Please test https://github.com/NixOS/nixpkgs/pull/254256

**Comment by StarLederer** [comment]
The binary is now present but unfortunately it is only able to fully work in vscode-fhs.

The binary itself works correctly but part of its behavior is to download and run vscode-server binaries from Microsoft (depending on which version of VS Code client it trying to connect) and those binaries are not patched. I'm not an expert at nix but as far as I remember there were tools that can wrap most binaries and make them work... nix-ld maybe? Would it be possible to apply a source patch to vscode to make it run server binaries inside that?

**Comment by bobby285271** [comment]
A quick try gave me message like `/home/bobby285271/.vscode/cli/servers/Stable-8b617bd08fd9e3fc94d14adb8d358b56e3f72314.staging/server/node: cannot execute: required file not found`, the related code is https://github.com/microsoft/vscode/commit/5cc83f79943d1de0abefa00d650a2533e30be99a.

I am thinking of something like https://github.com/NixOS/nixpkgs/blob/abc15c7f/pkgs/applications/editors/vscode/extensions/ms-vscode-remote.remote-ssh/default.nix#L18 ? I am not sure.

**Comment by StarLederer** [comment]
Yeah, I am getting the same error and I actually think I know what is causing it, and might have an idea how we can fix it.

Turns out code-server binary is not an ld binary and it is not what is causing the error but there is another known issue with with it which is the built-in node binary that comes with code-server. [This project](https://github.com/nix-community/nixos-vscode-server) attempts to fix this issue system-wide but we can't use it because it creates a system service to monitor paths where this node might get installed and intercepts that. The project is also designed primarily for [Dev Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) in which we have no control over the installing client and cannot patch it.

Luckily we don't have this limitation for tunnels which I think allows us to fix this in source.I looked through the file that you sent and I think something like the following might be worth a try.

![image](https://github.com/NixOS/nixpkgs/assets/55361752/8d9ff7b4-79ef-46f2-b243-d09e98576fd4)

I didn't run this code yet so no idea if it works or even compiles but I hope it can be helpful to you. I am assuming you know a bit of Rust and have time to do this yourself but let me know if you would like to try to get that code to run instead. I might be slow to respond but I am happy to help!

**Comment by bobby285271** [comment]
We are not building vscode from source right now I believe :woman_shrugging:

**Comment by StarLederer** [comment]
Oh, right... because Microsoft branded VS Code is proprietary... Well I'm not sure if this problem can be addressed than. I guess having it in vscode-fhs only is good enough for now
